spring-transaction/src/test/kotlin/org/jetbrains/exposed/spring/SpringTransactionManagerTest.kt

```diff
@@ -0,0 +1,466 @@
+        tm.isRollbackOnCommitFailure = true
+        assertFailsWith<TransactionSystemException> {
+            tm.executeAssert()
+        }
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "commit", "isClosed", "rollback", "close"))
+        assertEquals(1, con1.commitCallCount)
+        assertEquals(1, con1.rollbackCallCount)
+        assertEquals(1, con1.closeCallCount)
+    }
+
+    @Test
+    fun `transaction with exception on rollback`() {
+        con1.mockRollback = { throw SQLException("Rollback failure") }
+
+        val tm = SpringTransactionManager(ds1)
+        assertFailsWith<TransactionSystemException> {
+            tm.executeAssert {
+                assertEquals(false, it.isRollbackOnly)
+                it.setRollbackOnly()
+                assertEquals(true, it.isRollbackOnly)
+            }
+        }
+
+        assertTrue(con1.verifyCallOrder("setReadOnly", "setAutoCommit", "isClosed", "rollback", "close"))
+        assertEquals(1, con1.rollbackCallCount)
+        assertEquals(1, con1.closeCallCount)
+    }
+
+    private fun SpringTransactionManager.executeAssert(
+        initializeConnection: Boolean = true,
+        body: (TransactionStatus) -> Unit = {}
+    ) {
+        TransactionTemplate(this).executeWithoutResult {
+            assertEquals(
+                TransactionManager.managerFor(TransactionManager.currentOrNull()?.db),
+                TransactionManager.manager
+            )
+            if (initializeConnection) TransactionManager.current().connection
+            body(it)
+        }
+    }
+
+    class DataSourceSpy : DataSource {
+
+        private var connection: Connection? = null
+            set(value) {
+                field = value
+            }
+
+        fun clearMock() {
+            connection = null
+        }
+
+        fun mockConnection(connection: Connection) {
+            this.connection = connection
+        }
+
+        override fun getLogWriter(): PrintWriter {
+            throw NotImplementedError()
+        }
+
+        override fun setLogWriter(out: PrintWriter?) {
+            throw NotImplementedError()
+        }
+
+        override fun setLoginTimeout(seconds: Int) {
+            throw NotImplementedError()
+        }
+
+        override fun getLoginTimeout(): Int {
+            throw NotImplementedError()
+        }
+
+        override fun getParentLogger(): Logger {
+            throw NotImplementedError()
+        }
+
+        override fun <T : Any?> unwrap(iface: Class<T>?): T {
+            throw NotImplementedError()
+        }
+
+        override fun isWrapperFor(iface: Class<*>?): Boolean {
+            throw NotImplementedError()
+        }
+
+        override fun getConnection() = if (connection != null) connection!! else throw NotImplementedError()
+        override fun getConnection(username: String?, password: String?): Connection {
+            throw NotImplementedError()
+        }
+    }
+
+    class ConnectionSpy : Connection {
```

Comment:
There's quite a lot of overrides when implementing `Connection` from scratch.

I realize that the goal of this test suite is to mock transaction and connection calls without relying on an actual database or actual returned results. Would the validity or isolation of these unit tests be significantly compromised if the data source at least started with a base connection from the H2 driver? Especially if all the function calls being tested are overriden in `ConnectionSpy`?

Please think about it and consider the following alternative:
```kt
class ConnectionSpy(private val connection: Connection) : Connection by connection {
    // then remove all overrides not implemented
}

class DataSourceSpy(private val connectionSpy: (Connection) -> Connection) : DataSource {
    // ...
    override fun getConnection() = connection ?: run {
        val testConnection = DriverManager.getConnection("jdbc:h2:mem:test")
        connectionSpy(testConnection)
    }
}

private val ds1 = DataSourceSpy(::ConnectionSpy)
private val con1 = ConnectionSpy(ds1.connection)
// `DriverManager.getConnection()` would only be invoked once on this first access
// then the value would be accessed from the private variable `connection`
// which can continue to be stored using `mockConnection()`
// or by adding an assignment in the null branch of `getConnection()` above.
```

Otherwise maybe the 2 spy classes need to at least be extracted to their own file and I'll get a second opinion again on what our options are.

Reply:
I used the neck object to test the behavior of the connection object and if there is another way to do it, I will use it. If it makes sense to test whether or not the method is called in the way you suggested using the example code above, I would change it to do so.

Reply:
```
    private open class DataSourceSpy(private val connectionSpy: (Connection) -> Connection) : DataSource {
        var con: Connection = connectionSpy(DriverManager.getConnection("jdbc:h2:mem:test"))

        override fun getConnection() = con
        // other override functions that throw not implement exception
    }

    private open class ConnectionSpy(private val connection: Connection) : Connection by connection {

        var commitCallCount: Int = 0
        ...
        // mock functions that need to test
     }
``` 

I changed it to use the H2 connection implementation as you suggested.

Since the purpose is to check whether the method is called or not, I think this should be able to complete the test without any problems. If you think there is a better way to do this, or if you think I should move the Spy class to a different file, please comment.